### PR TITLE
[core] Handle CoreWorker with multiple threads

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?
Right now, `CoreWorker` assumes each thread for a  core worker process will execute task. It uses a thread_local `WorkerThreadContext` to capture the running task on a single thread. However, this assumption doesn't hold for at least the below scenarios: 
1. A task (normal task/actor) not as a threaded actor/async actor that runs some ray code (submitting another code, getting runtime_context) in another thread. 
2. An asyncio actor, which will execute the async actor method in a thread that is not known to ray's runtime. 

With the PR, it has the below behavior changes:
1. When a task runs another thread that runs ray code, not on an async actor or a threaded actor:
   - Before: the new thread will have a thread context with random task id
   - Now: the new thread will have the same task id as the original task. 

2. An asyncio actor method:
   - Before: the method might run in a thread that has random task id
   - Now: the new thread will yield a Nil task id since when the concurrency allowed is greater than 1, there is no way for Ray to figure out which task the user-level thread is actually running. 

#### On a high level, this PR 
- Removes the thread local thread context, but renamed it as `WorkerExecContext`. 
- In each `WorkerContext`, it keeps a mapping of thread id to `WorkerExecContext`. For actors with out of order execution (i.e., threaded actor), this map is used to retrieve the executing task runs by each thread. 
- When running `ExecTask` in the coreworker, the `WorkerExecContext` will be set properly with the task spec and task id. 
- Before running a task or after running a task, the `WorkerExecContext` will have Nil task id info
- Adds a `current_exec_context` to the `WorkerContext` to capture most recently executed task's context, this allows us to return the executing task info (its task id) when the task spawns new thread in user space (when it's not an async actor/threaded actor). 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - Running the script https://gist.github.com/rickyyx/61026bdf5110f5a56d840f9a04debec7 yields no random parent task id
   - [x] Unit tests

### CODE REVIEW

File: src/ray/core_worker/context.h [hunk 1]
```diff
@@ -129,15 +139,15 @@ class WorkerContext {
   // The runtime env info.
   std::shared_ptr<rpc::RuntimeEnvInfo> runtime_env_info_ GUARDED_BY(mutex_);
   /// The id of the (main) thread that constructed this worker context.
-  const boost::thread::id main_thread_id_;
+  const std::thread::id main_thread_id_ GUARDED_BY(mutex_);
```
Comment (rickyyx) [inline src/ray/core_worker/context.h:142]: `boost:thread:id` is not hashable, and to be used as key for absl::flat_hash_map somehow....

File: python/ray/tests/test_runtime_context.py [hunk 2]
```diff
@@ -338,6 +338,90 @@ def test_errors_when_ray_not_initialized():
...
+def test_async_actor_runtime_context(ray_start_regular):
+    @ray.remote
+    class AsyncActor:
+        async def task_id(self):
+            return ray.get_runtime_context().get_task_id()
+
+    # Asyncio actors run in another thread that's beyond knowledge of ray
+    async_actor = AsyncActor.options(max_concurrency=1).remote()
+    tasks = [async_actor.task_id.remote() for _ in range(10)]
+    expect_task_ids = [None] * len(tasks)
```
Comment (rickyyx) [inline python/ray/tests/test_runtime_context.py:364]: So, is this the right behavior we should have for async actor now? 

If `max_concurrency = 1`, we could actually return `current_exec_context` to make it work. But if `max_concurrency > 1`, we have no way to know which task the thread is executing.

File: src/ray/core_worker/context.cc [hunk 3]
```diff
@@ -264,8 +260,9 @@ void WorkerContext::SetCurrentActorId(const ActorID &actor_id) LOCKS_EXCLUDED(mu
 void WorkerContext::SetTaskDepth(int64_t depth) { task_depth_ = depth; }
 
 void WorkerContext::SetCurrentTask(const TaskSpecification &task_spec) {
-  GetThreadContext().SetCurrentTask(task_spec);
   absl::WriterMutexLock lock(&mutex_);
+  InitExecContext(task_spec.TaskId());
+  current_exec_context_->SetCurrentTask(task_spec);
```
Comment (scv119) [inline src/ray/core_worker/context.cc:265]: should we call `GetExecContextInternal` ?

**Review by scv119 — CHANGES_REQUESTED** [review]

File: src/ray/core_worker/context.h [hunk 4]
```diff
@@ -119,6 +126,9 @@ class WorkerContext {
   ActorID current_actor_id_ GUARDED_BY(mutex_);
   int current_actor_max_concurrency_ GUARDED_BY(mutex_) = 1;
   bool current_actor_is_asyncio_ GUARDED_BY(mutex_) = false;
+  // If this is an actor with multiple threads of execution. It could be an async actor or
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply src/ray/core_worker/context.h:129]: Oh, I guess if we use `MaxActorConcurrency` as you suggested, then non concurrent async actor could be handled, which is single threaded. The only thing is it uses a user-spawned thread for execution.

File: src/ray/core_worker/context.cc [hunk 5]
```diff
@@ -350,15 +349,38 @@ bool WorkerContext::CurrentActorDetached() const {
...
+           "with nil task id.";
+    return std::make_shared<WorkerExecContext>(TaskID::Nil());
+  }
+
+  return itr->second;
+}
+
+WorkerExecContext &WorkerContext::GetExecContext() const {
+  absl::ReaderMutexLock lock(&mutex_);
+  return *GetExecContextInternal();
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply src/ray/core_worker/context.cc:382]: ah right - I probably need to change the interface a bit. 

What do you think about the behavior when we couldn't find a thread's corresponding ctx, and it has concurrency > 1 (e.g. a multi threaded actor starts a user level thread). Is a dummy context (Nil task id) appropriate?

File: src/ray/core_worker/context.cc (hunk 3 above)
Comment (rickyyx) [reply src/ray/core_worker/context.cc:265]: We could - it's kind of implicit, where after initing the context, `current_exec_context_` will also point to the current ctx (same thread) with the writer lock acquired. 
So it's still correct but I guess calling `GetExecContextInternal` will be more readable as you pointed out.

File: src/ray/core_worker/context.cc [hunk 6]
```diff
@@ -278,6 +275,7 @@ void WorkerContext::SetCurrentTask(const TaskSpecification &task_spec) {
     current_actor_is_direct_call_ = true;
     current_actor_max_concurrency_ = task_spec.MaxActorConcurrency();
     current_actor_is_asyncio_ = task_spec.IsAsyncioActor();
+    current_actor_is_multi_threaded_ = task_spec.ExecuteOutOfOrder();
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply src/ray/core_worker/context.cc:278]: oh, that makes sense.

File: python/ray/tests/test_runtime_context.py [hunk 7]
```diff
@@ -338,6 +338,90 @@ def test_errors_when_ray_not_initialized():
...
+    submitted_task_ids = [task.task_id().hex() for task in tasks]
+    actual_task_ids = ray.get(tasks)
+    assert ",".join(submitted_task_ids) == ",".join(actual_task_ids)
+
+
+def test_async_actor_runtime_context(ray_start_regular):
+    @ray.remote
+    class AsyncActor:
+        async def task_id(self):
+            return ray.get_runtime_context().get_task_id()
```
Comment (alanwguo) [inline python/ray/tests/test_runtime_context.py:359]: what about the case where an async actor method calls another ray task. Will the child ray task know the parent is the actor_task? 
```
    class AsyncActor:
        async def method(self):
            return await foo.remote()
```

will the foo task instance know that AsyncActor.method is its parent?

Comment (rickyyx) [reply python/ray/tests/test_runtime_context.py:359]: yep - good question. adding test cases to explicitly cover it.

File: src/ray/core_worker/context.h [hunk 8]
```diff
@@ -129,15 +139,15 @@ class WorkerContext {
   // The runtime env info.
   std::shared_ptr<rpc::RuntimeEnvInfo> runtime_env_info_ GUARDED_BY(mutex_);
   /// The id of the (main) thread that constructed this worker context.
-  const boost::thread::id main_thread_id_;
+  const std::thread::id main_thread_id_ GUARDED_BY(mutex_);
+  // All thread contexts started by this CoreWorker.
+  absl::flat_hash_map<std::thread::id, std::shared_ptr<WorkerExecContext>>
+      all_exec_threads_contexts_ GUARDED_BY(mutex_);
+  // The current running task's thread context.
+  std::shared_ptr<WorkerExecContext> current_exec_context_ GUARDED_BY(mutex_);
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply src/ray/core_worker/context.h:150]: it's not only for actor, normal task also refers to this.

**Pull request closed without merging**